feat: Add Copilot agent instructions (migrated from OMEGA) (#35322)
openvinotoolkit/openvino: .github/agents-prototype/skills/add-fusion-transformation/step1-analysis.md

# Skill: OpenVINO Transformation Analysis

## Purpose

Analyse an OpenVINO IR sub-graph to identify the optimal fusion pattern and
transformation type before any implementation work begins. Produces a structured
`transformation_analysis.md` that the implementation skill consumes.

## When to invoke

- Transformation Agent Step 1
- Any time you need to decide *which* `ov::pass` class to use and *why*

---

## Steps

### Step 1: Load Op Spec and Existing Graph

Obtain the op spec from agent-results:
```bash
# Op spec is written by Core OpSpec Agent to agent-results/core-opspec/
OP_SPEC_PATH=$(python3 -c "
import json
try:
    d = json.load(open('agent-results/core-opspec/core_opspec_result.json'))
    print(d.get('op_spec_path', ''))
except Exception:
    print('')
")
[ -n "$OP_SPEC_PATH" ] && cat "$OP_SPEC_PATH" || echo "[WARN] Op spec not found in agent-results/core-opspec/"

> ```

Reviewer: To retrive the op spec - is the python code snippet really needed? Are the hardcoded paths ensured to be always correct? 

My recommendation is to ask to obtain the op spec from the common docs directory (there can be already existing, moreover any agent should write the spec directly to the openvino docs folder as welli).

```suggestion
Obtain the op spec from the openvino docs.
```

Export the model to IR (using cached IR if available) and visualise the target
sub-graph:
```python
import openvino as ov

model = ov.Core().read_model("openvino_model.xml")
# Find the target nodes by op type or friendly name
target = [n for n in model.get_ordered_ops()
          if n.get_type_name() in {"MatMul", "Add", "Tanh"}]
for n in target:
    print(n.get_friendly_name(), n.get_type_name(),
          [i.get_shape() for i in n.inputs()])

> ```

Reviewer: It's not the code to be executed, it should be mentioned that it's an example for reference.
```suggestion
sub-graph, example:
```

### Step 2: Draw the Sub-Graph Pattern

> Produce a textual node diagram:

Reviewer: It's not the code to be executed, it should be mentioned that it's an example for reference.
```suggestion
sub-graph, example:
```

Apply this decision tree:

```
Is the fusion local (fixed sub-graph topology)?
  YES → Does it need to run before any other pass touches these nodes?
    YES → MatcherPass registered early in common_optimizations
    NO  → MatcherPass registered at the appropriate priority slot
  NO → Does the pass need consumer information (backward)?
    YES → BackwardGraphRewrite
    NO  → FunctionPass (iterate over all nodes manually)
```

| Type | When | Examples in codebase |
|---|---|---|
| `MatcherPass` | Local topology, pattern of 2-10 ops | `FuseMHATokensSplit`, `FuseConvolutionSimpleLeakyRelu` |
| `FunctionPass` | Multi-pattern or graph-wide | `ConstantFolding`, `AlignMixedFP32FP16Types` |
| `BackwardGraphRewrite` | Consumer-first needed | `ConvertPrecision` (when checking all consumers) |
| `GraphRewrite` | Forward multi-pattern | Wraps multiple `MatcherPass` instances |

### Step 5: Find a Template to Follow

Search for the syntactically closest existing transformation:
```bash
# Find transformations that match a similar pattern (e.g., MatMul + Add fusion)
grep -r "MatMul\|GemmFusion\|LinearFusion" \
  src/common/transformations/include/transformations/ --include="*.hpp" -l
```

> Read that file fully — it is your implementation template.

Reviewer: Instructions like that are the direction I mentioned in previous comment, a suggestion what to look for, but without hardcoded grep, the agents can perform better if they can deduce what "similar pattern" means, while hardcoded MatMul here can be misleading.

```suggestion
### Step 5: Find a Template to Follow

Search for the syntactically closest existing transformation.
Read that file — it is your implementation template.
```
